ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckManager.java

```diff
@@ -0,0 +1,76 @@
+package com.github.ambry.replication;
+
+import com.codahale.metrics.MetricRegistry;
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.clustermap.ClusterParticipant;
+import com.github.ambry.clustermap.DataNodeId;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.config.ReplicationConfig;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.network.ConnectionPool;
+import com.github.ambry.notification.NotificationSystem;
+import com.github.ambry.server.StoreManager;
+import com.github.ambry.store.MessageInfo;
+import com.github.ambry.store.StoreKeyConverter;
+import com.github.ambry.store.StoreKeyConverterFactory;
+import com.github.ambry.store.StoreKeyFactory;
+import com.github.ambry.store.Transformer;
+import com.github.ambry.utils.Utils;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+
+
+public class BackupCheckManager extends ReplicationManager {
+  public BackupCheckManager(ReplicationConfig replicationConfig, ClusterMapConfig clusterMapConfig,
+      StoreConfig storeConfig, StoreManager storeManager, StoreKeyFactory storeKeyFactory, ClusterMap clusterMap,
+      ScheduledExecutorService scheduler, DataNodeId dataNode, ConnectionPool connectionPool,
+      MetricRegistry metricRegistry, NotificationSystem requestNotification,
+      StoreKeyConverterFactory storeKeyConverterFactory, String transformerClassName,
+      ClusterParticipant clusterParticipant, Predicate<MessageInfo> skipPredicate) throws ReplicationException {
+    super(replicationConfig, clusterMapConfig, storeConfig, storeManager, storeKeyFactory, clusterMap, scheduler,
+        dataNode, connectionPool, metricRegistry, requestNotification, storeKeyConverterFactory, transformerClassName,
+        clusterParticipant, skipPredicate);
+  }
+
+  /**
+   * Create thread pool for a datacenter.
+   * @param datacenter The datacenter String.
+   * @param numberOfThreads Number of threads to create for the thread pool.
+   * @param startThread If thread needs to be started when create.
+   */
+  protected List<ReplicaThread> createThreadPool(String datacenter, int numberOfThreads, boolean startThread) {
+    nextReplicaThreadIndexByDc.put(datacenter, new AtomicInteger(0));
```

Comment:
There is a lot of duplicate code here. The only difference between `BackupCheckManager::createThreadPool` and `ReplicationManager:: createThreadPool` is that one uses `ReplicaThread` while the other uses `BackupCheckerThread`. We can create a method like `getReplicaThreadObj` in ReplicationManager and override it in BackupCheckManager. That will do it.

Reply:
good idea!

Reply:
I have looked into this and doing this makes the PR bigger than it is and possibly hard to review.
Reason is this.

5 call sites -> addRemoteReplicaInfoToReplicaThread -> getOrCreateThreadPoolIfNecessary -> createThreadPool

ReplicaConfig is not passed in anywhere. So I'll have to change all these function calls all the way to the top to pass in ReplicaConfig or any other Config to know what type of threadObj to create. Creating a new manager in this context is cleaner.

Reply:
Actually hold on, I think i figured it out